### PR TITLE
Implement NoCommonSubexpressionElimination

### PR DESCRIPTION
**Pull request by sequencer**
### Contributor Checklist

- [x] Did you add Scaladoc to every public function/method?
- [ ] Did you update the FIRRTL spec to include every new feature/behavior?
- [ ] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [x] Did you state the API impact?
- [x] Did you specify the code generation impact?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement
- new feature/API

#### API Impact
Add `--no-common-subexpression-elimination` to disable CSE.

#### Backend Code Generation Impact
None

#### Desired Merge Strategy
- Rebase: You will rebase the PR onto master and it will be merged with a merge commit.

#### Release Notes
Add `--no-common-subexpression-elimination` to disable CSE.

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels?
- [ ] Did you mark the proper milestone (1.2.x, 1.3.0, 1.4.0) ?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you mark as `Please Merge`?

### CODE REVIEW

**Comment by sequencer** [comment]
Asking for review from @seldridge.

File: src/main/scala/firrtl/passes/CommonSubexpressionElimination.scala [hunk 1]
```diff
@@ -2,19 +2,31 @@
...
 
   override def prerequisites = firrtl.stage.Forms.LowForm
   override def optionalPrerequisiteOf =
     Seq(Dependency[SystemVerilogEmitter], Dependency[VerilogEmitter])
 
   override def invalidates(a: Transform) = false
 
+  val options = Seq(
+    new ShellOption[Unit](
+      longOption = "no-common-subexpression-elimination",
```
Comment (seldridge) [inline src/main/scala/firrtl/passes/CommonSubexpressionElimination.scala:24]: It's probably fine to keep the option shorter:

```suggestion
      longOption = "no-cse",
```

**Review by seldridge — APPROVED** [review]
Looks great.  It makes sense to me to have more fine-grained control over which optimizations are run.

**Comment by seldridge** [comment]
I marked this as 1.5 due to perceived binary incompatibility issues around the change to the inheritance/mix-ins for CSE. (I think this is the case?) This patch be backported with a different approach, though I'm not sure this is necessary to backport. @sequencer: if you need this in 1.4.x or earlier, feel free to update. 😄

**Comment by sequencer** [comment]
@seldridge 1.5 is fine :) let's put it into 1.5.

**Comment by lingscale** [comment]
Just like NoDCEAnnotation and NoConstantPropagationAnnotation, maybe NoCommonSubexpressionEliminationAnnotation more perfect than NoCommonSubexpressionElimination?

**Pull request merged**